src/main/kotlin/org/rust/lang/core/psi/util/RustAttrExtensions.kt

```diff
@@ -0,0 +1,65 @@
+package org.rust.lang.core.psi.util
+
+import org.rust.lang.core.psi.*
+
+/**
+ * Appends a new meta item `new` to an attribute with a list of arguments:
+ *
+ * ```
+ * #[attr]              ->  #[attr(new)]
+ * #[attr="something"]  ->  #[attr(new)]
+ * #[attr(something)]   ->  #[attr(something, new)]
+ * #[attr(something]    ->  #[attr(something, new]
+ * #[attr()]            ->  #[attr(new)]
+ * #[attr(]             ->  #[attr(new]
+ * ```
+ *
+ * @return `meta` argument or its copy if insertion was successful, null otherwise.
+ */
+fun RustAttr.appendMetaToList(meta: RustMetaItem): RustMetaItem? {
+    // extract attribute internals
+    val attrMeta = when (this) {
+        is RustInnerAttr -> metaItem
+        is RustOuterAttr -> metaItem
+        else             -> return null
+    } ?: return null
+
+    if (attrMeta.lparen == null && attrMeta.rparen == null) {
+        // if both parens are missing, remove everything inside the attribute except its identifier and add `meta`
+        // between parens to it
+        if (attrMeta.identifier.nextSibling != null) {
+            attrMeta.deleteChildRange(attrMeta.identifier.nextSibling, attrMeta.lastChild)
+        }
+
+        val tempMeta = RustElementFactory.createMeta(this.project, "whatever(item)") ?: return null
+        tempMeta.metaItem!!.replace(meta)
+
+        val newLeftParen = attrMeta.addRangeAfter(tempMeta.lparen, tempMeta.rparen, attrMeta.identifier)
+        return newLeftParen.nextSibling as RustMetaItem
+
+    } else if (attrMeta.lparen != null) {
+        // if left paren is present, add `meta` to the end of the list of items
+        val leftParen = attrMeta.lparen!!
+
+        if (leftParen.nextSibling != null && leftParen.nextSibling != attrMeta.rparen) {
+            // if there are other meta items in parens, add `meta` after the last one of them
+            val elementToInsertAfter = if (attrMeta.rparen != null)
+                attrMeta.rparen!!.prevSibling!!
+            else
+                attrMeta.lastChild!!
+
+            // we also need a comma, so first we insert a comma and a temporary meta item from another well-formed meta
+            val tempMeta = RustElementFactory.createMeta(this.project, "whatever(item,item)") ?: return null
+            val start = tempMeta.lparen!!.nextSibling!!.nextSibling!!
```

Comment:
>and I will certainly try to understand the logic and make sure that it is still correct after refactoring.  I hope this is true for others as well.

Humans (well, at least me :) ) tend to make mistakes (especially when it comes to logic), so imo it is much better to turn assertions into compile time errors. I think that in this particular case introducing two variables, `leftParen` and `rightParen` would substantially reduce the need in `!!`. And this does not need any additional ifs, they are already there (but Kotlin can't deduce non nulability for `foo.bar` when `bar` is a getter, hence the need in local `val`s). 

The right way to turn this into compile time errors would be to make the following PSI (which I might do some day)

```
interface RustMetaItem {
    @Nullable MetaItemArgList getArgList()
}

interface MetaItemArgList {
    @NotNull PsiElement getLparen()
    @NotNull List<RustMetaItem> getArgs()
    @NotNull PsiElement getRparen()
}
```

However, I think we can completely sidestep the issue if instead of custom logic for adding meta item to arbitrary attr, we will just do the following: 

```
        val deriveAttr = if (existingDeriveAttr != null) existingDeriveAttr
        else {
            val attr = RustElementFactory.createOuterAttr(project, "derive()") ?: return
            when (item) {
                is RustStructItem -> item.addBefore(attr, item.struct)
                is RustEnumItem   -> item.addBefore(attr, item.enum)
                else              -> return
            } as RustOuterAttr
        }
        // Just move the caret and don't insert anything.
        val offset = deriveAttr.metaItem?.rparen?.textOffset ?: deriveAttr.textOffset
        editor?.caretModel?.moveToOffset(offset)
        return
```

It is not as fancy as with adding a coma (btw, have you handled the case when there is already a trailing coma in the arg list?), but hopefully much more robust. What do you think?